pkg/kube/krt/files.go

```diff
@@ -0,0 +1,107 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"go.uber.org/atomic"
+	"sigs.k8s.io/yaml"
+
+	"istio.io/istio/pkg/filewatcher"
+	"istio.io/istio/pkg/ptr"
+)
+
+type FileCollection[T any] struct {
+	StaticCollection[T]
+}
+
+func NewFileCollection[T any](opts ...CollectionOption) FileCollection[T] {
+	panic("not yet implemented")
+}
+
+type FileSingleton[T any] struct {
+	Singleton[T]
+}
+
+// NewFileSingleton returns a collection that reads and watches a single file
+// The `readFile` function is used to read and deserialize the file and will be called each time the file changes.
+// This will also be called during the initial construction of the collection; if the initial readFile fails an error is returned.
+func NewFileSingleton[T any](
+	fileWatcher filewatcher.FileWatcher,
+	filename string,
+	readFile func(filename string) (T, error),
+	opts ...CollectionOption,
+) (FileSingleton[T], error) {
+	cfg, err := readFile(filename)
+	if err != nil {
+		return FileSingleton[T]{}, err
```

Comment:
Do we want it to be terminal if the first readFile returns an error? It seems like we could still setup a watch and the file may become valid later

Reply:
I agree the behavior is a little weird. I think we probably should do what you said, assuming the file watcher handles it (not sure the details of how it behaves, watches on files are a bit wonky in fsnotify)... however, I may want to split this out. With a big refactoring like this is like to keep the behavior as close as possible to before so we can isolate refactoring vs behavioral changes (like https://github.com/istio/istio/issues/54615 would probably be something we followup as well)

Reply:
+1 to changing the behavior in a follow-up

Reply:
Sounds good. I'm OK with followup